cpp/src/arrow/ipc/reader.cc

```diff
@@ -860,27 +861,40 @@ Status ReadTensor(const Message& message, std::shared_ptr<Tensor>* out) {
 
 namespace {
 
-Status ReadSparseCOOIndex(const flatbuf::SparseTensor* sparse_tensor, int64_t ndim,
-                          int64_t non_zero_length, io::RandomAccessFile* file,
-                          std::shared_ptr<SparseIndex>* out) {
+Status ReadSparseCOOIndex(const flatbuf::SparseTensor* sparse_tensor,
+                          const std::vector<int64_t>& shape, int64_t non_zero_length,
+                          io::RandomAccessFile* file, std::shared_ptr<SparseIndex>* out) {
   auto* sparse_index = sparse_tensor->sparseIndex_as_SparseTensorIndexCOO();
+
+  std::shared_ptr<DataType> indices_type;
+  RETURN_NOT_OK(internal::GetSparseCOOIndexMetadata(sparse_index, &indices_type));
+
   auto* indices_buffer = sparse_index->indicesBuffer();
   std::shared_ptr<Buffer> indices_data;
   RETURN_NOT_OK(
       file->ReadAt(indices_buffer->offset(), indices_buffer->length(), &indices_data));
-  std::vector<int64_t> shape({non_zero_length, ndim});
-  const int64_t elsize = sizeof(int64_t);
-  std::vector<int64_t> strides({elsize, elsize * non_zero_length});
+  std::vector<int64_t> indices_shape(
+      {non_zero_length, static_cast<int64_t>(shape.size())});
+  auto* indices_strides = sparse_index->indicesStrides();
+  std::vector<int64_t> strides;
+  if (indices_strides->size() > 0) {
```

Comment:
What if `indices_strides->size() != 2`? Shouldn't we validate input here?

Reply:
Indeed.  This check should be done in the constructor of `SparseCOOIndex` class.  I'll fix it.